subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultArtifactTransformOutputs.java

```diff
@@ -65,18 +69,28 @@ public static void validateOutputExists(File output) {
         ImmutableList<File> outputs = outputsBuilder.build();
         for (File output : outputs) {
             validateOutputExists(output);
+            if (outputFiles.contains(output) && !output.isFile()) {
+                throw new InvalidUserDataException("Transform output file " + output.getPath() + " must be a file, but is not.");
+            }
+            if (outputDirectories.contains(output) && !output.isDirectory()) {
+                throw new InvalidUserDataException("Transform output directory " + output.getPath() + " must be a directory, but is not.");
+            }
         }
         return outputs;
     }
 
     @Override
     public File dir(Object path) {
-        return resolveAndRegister(path);
+        File outputDir = resolveAndRegister(path);
+        outputDirectories.add(outputDir);
+        return outputDir;
     }
 
     @Override
     public File file(Object path) {
-        return resolveAndRegister(path);
+        File outputFile = resolveAndRegister(path);
+        outputFiles.add(outputFile);
+        return outputFile;
```

Comment:
Would it take anything else than a call to `GFileUtils.mkdirs()` in these methods to ensure that the directories are there? If it's that simple, I think we should do it.

Reply:
Yes, we need to check if the file refers a path in the input artifact or in the output. It is not hard, I can add it.

Reply:
Creating the directories now.